### PR TITLE
Improvements and parallelism for cf-net

### PR DESCRIPTION
**Pull request by vpodzime**
The first few commits are just minor fix-ups and refactoring for *cf-net*. The last two ones add support for getting files in parallel (in threads). I tried playing with the number of threads and everything up to ~250 was working fine for me.

### CODE REVIEW

**Comment by olehermanse** [comment]
@vpodzime I will review this when release work is done

File: cf-net/cf-net.c [hunk 1]
```diff
@@ -704,6 +721,17 @@ static int CFNetGet(ARG_UNUSED CFNetOptions *opts, const char *hostname, char **
                 specified_path = true;
                 break;
             }
+            case 'j':
+            {
+                char *tail = NULL;
+                n_threads = (int) strtol(optarg, &tail, 10);
```
Comment (jimis) [inline cf-net/cf-net.c]: I believe @olehermanse has a new function for this purpose.

File: cf-net/cf-net.c [hunk 2]
```diff
@@ -704,6 +721,17 @@ static int CFNetGet(ARG_UNUSED CFNetOptions *opts, const char *hostname, char **
...
+                if (tail && (*tail != '\0'))
+                {
+                    printf("Failed to parse number of threads/jobs: %m");
```
Comment (jimis) [inline cf-net/cf-net.c]: forgot the terminating LF. ;-)

File: cf-net/cf-net.c [hunk 3]
```diff
@@ -740,9 +768,62 @@ static int CFNetGet(ARG_UNUSED CFNetOptions *opts, const char *hostname, char **
         return -1;
     }
 
-    int ret = CFNetGetFile(hostname, remote_file, local_file);
+    CFNetThreadData **threads = (CFNetThreadData**) xcalloc(n_threads, sizeof(CFNetThreadData*));
+    for (int i = 0; i < n_threads; i++)
+    {
+        threads[i] = (CFNetThreadData*) xcalloc(1, sizeof(CFNetThreadData));
+        threads[i]->data = (GetFileData*) xcalloc(1, sizeof(GetFileData));
```
Comment (jimis) [inline cf-net/cf-net.c]: no casting is needed nor recommended in all of the above, in C void* casting is implicit.

File: cf-net/cf-net.c [hunk 4]
```diff
@@ -740,9 +768,62 @@ static int CFNetGet(ARG_UNUSED CFNetOptions *opts, const char *hostname, char **
...
+        threads[i]->data->hostname = hostname;
+        threads[i]->data->remote_file = remote_file;
+        if (n_threads > 1)
+        {
+            if (strstr(local_file, "%d") != NULL)
+            {
+                snprintf(threads[i]->data->local_file, PATH_MAX, local_file, i);
```
Comment (jimis) [inline cf-net/cf-net.c]: dangerous but I guess it's ok for a testing tool

File: cf-net/cf-net.c (hunk 1 above)
Comment (olehermanse) [reply cf-net/cf-net.c]: @vpodzime @jimis `StringToLongExitOnError()` from `string_lib.h` would be appropriate. `StringToLong()` if you want to do custom error handling. (I plan to do some changes, like actually returning long on platforms where `long != int`, but they will not be breaking.)

File: cf-net/cf-net.c (hunk 4 above)
Comment (vpodzime) [reply cf-net/cf-net.c]: Yeah, there's practically no chance that ``local_file`` will not be ``\0``-terminated.

Comment (jimis) [reply cf-net/cf-net.c]: it's not for the NULL termination, it's for the possibility of putting format string qualifiers in the filename. Imagine a filename "blah%s%x%d".

Comment (vpodzime) [reply cf-net/cf-net.c]: Oh, that's right. Well, we are all responsible people not doing things like this. :)

**Comment by jimis** [comment]
Acceptance test or even a load test (in `tests/load`). But since this is a testing tool for now, I am lenient as to a test exists or not.

**Comment by vpodzime** [comment]
> But since this is a testing tool for now, I am lenient as to a test exists or not.

IMHO the best way to test a test tool is to run tests with it.

**Comment by jimis** [comment]
> IMHO the best way to test a test tool is to run tests with it.

Well that's your main task AFAIK, but feel free to include a small load test in Core, the more the better!

**Comment by olehermanse** [comment]
@vpodzime Please tell me if you want this merged now.

**Comment by vpodzime** [comment]
> @vpodzime Please tell me if you want this merged now.

That would be great. I plan to add some core-limited tests based on this, but that can happen in a separate PR once I get back to it. Thanks @olehermanse!

**Pull request merged**